src/neo4j_genai/pipeline/pipeline.py

```diff
@@ -0,0 +1,470 @@
+
+    def reinitialize(self) -> None:
+        self.status = RunStatus.SCHEDULED
+
+    async def run(
+        self, data: dict[str, Any], callback: Callable[[Any, Any, Any], Awaitable[Any]]
+    ) -> None:
+        """Main method to execute the task.
+        1. Get the input defs (path to all required inputs)
+        2. Build the input dict from the previous input defs
+        3. Call the execute method
+        4. Call the pipeline callback method to deal with child dependencies
+        """
+        logger.debug(f"TASK START {self.name=} {data=}")
+        # prepare the inputs defs that needs to be
+        # propagated from parent components
+        input_config = await self.get_input_config_from_parents()
+        # create component's input based on initial input data,
+        # already done component's outputs and inputs definition mapping
+        inputs = self._pipeline.get_component_inputs(self.name, input_config, data)
+        # execute component task
+        res = await self.execute(**inputs)
+        if res is None:
+            return
+        logger.debug(f"TASK RESULT {self.name=} {res=}")
+        # save its results
+        await callback(self, data, res)
+
+
+class Orchestrator:
+    """Orchestrate a pipeline
+    Once a TaskNode is done, it calls the `on_task_complete` callback
+    that will save the results, find the next tasks to be executed
+    (checking that all dependencies are met), and run them.
+    """
+
+    def __init__(self, pipeline: "Pipeline"):
+        self.pipeline = pipeline
+
+    async def run_task(self, task: TaskPipelineNode, data: dict[str, Any]) -> None:
+        await task.run(data, callback=self.on_task_complete)
+
+    async def on_task_complete(
+        self, task: TaskPipelineNode, data: dict[str, Any], res: RunResult
+    ) -> None:
+        """When a given task is complete, it will call this method
+        to find the next tasks to run.
+        """
+        # first call the method for the pipeline
+        # this is where the results can be saved
+        self.pipeline.on_task_complete(task, res)
+        # then get the next tasks to be executed
+        # and run them in //
+        await asyncio.gather(*[self.run_task(n, data) async for n in self.next(task)])
+
+    async def check_dependencies_complete(self, task: TaskPipelineNode) -> None:
+        """Check that all parent tasks are complete.
+
+        Raises:
+            MissingDependencyError if a parent task's status is different from DONE.
+        """
+        dependencies = self.pipeline.previous_edges(task.name)
+        for d in dependencies:
+            start_node = self.pipeline.get_node_by_name(d.start)
+            d_status = await start_node.read_status()
+            if d_status != RunStatus.DONE:
+                logger.warning(
+                    f"Missing dependency {d.start} for {task.name} (status: {d_status})"
+                )
+                raise PipelineMissingDependencyError()
```

Comment:
Apologies if I'm misunderstanding something here but is it possible to run a chain of components in parallel and then 'gather' the results in some final component, or can parallel processing only be done within a component? e.g. after chunking would it be run a pipeline of components on each chunk, and only gather them together at the end for entity resolution? Or must the 'gathering' be done at the end of every component

Reply:
For now, parallelism is possible only within a component. The only "parallelism" the pipeline allows is between branches: for instance, the schema extractor and document chunker are run in parallel because they do not depend on each other.

Reply:
Yes I agree it makes sense to execute full pipelines in parallel for different chunks, but I'm not sure this package is the best place for it. I'd rather build the pipeline for a single chunk, and then rely on (potentially) distributed task managers (such as luigi?) to manage the workload. 
But if we decide to implement it ourselves, we'll have to make some changes to the way the pipelines currently work, but that should be doable.